src/AirplaneSecondary.java

```diff
@@ -0,0 +1,47 @@
+/**
+ * The secondary methods for AirplaneKernel, add the method bodies to finish the
+ * abstract class assignment.
+ */
+public abstract class AirplaneSecondary implements Airplane {
+
+    @Override
+    public final String displayDefinedMetrics() {
+        String output = "";
+        for (Mode mode : Mode.values()) {
+            if (this.isDefined(mode)) {
+                output += this.getMetric(mode);
+            }
+        }
+        return output;
+    }
+
+    @Override
+    public final String toString() {
+        return "Airplane [" + "Altitude=" + Mode.ALTITUDE + "," + "Direction="
```

Comment:
This won't work as expected. You'll need to call this.getMetric:
```suggestion
        return "Airplane [" + "Altitude=" + this.getMetric(Mode.ALTITUDE) + "," + "Direction="
```